dstep/translator/HeaderIndex.d

```diff
@@ -0,0 +1,267 @@
+    private string[size_t] reverse_;
+    private Set!(Tuple!(size_t, size_t)) directInclusions;
+    private Set!(Tuple!(size_t, size_t)) indirectInclusions;
+    private string internalPrefix;
+
+    public this(TranslationUnit translationUnit)
+    {
+        import std.algorithm;
+
+        auto directives = translationUnit.cursor.children
+            .filter!(cursor => cursor.kind == CXCursorKind.inclusionDirective);
+
+        this(directives);
+    }
+
+    public this(T)(T directives)
+    {
+        import std.random;
+        import std.range;
+
+        internalPrefix = generate(() => uniform!("[]")('a', 'z'))
+            .take(32).array.idup;
+
+        foreach (directive; directives)
+        {
+            auto path = directive.file.absolutePath;
+            auto includedPath = directive.includedFile.absolutePath;
+
+            if (path != "" && includedPath != "")
+            {
+                if (path !in headers_)
+                {
+                    reverse_[headers_.length] = path;
+                    headers_[path] = headers_.length;
+                }
+
+                if (includedPath !in headers_)
+                {
+                    reverse_[headers_.length] = includedPath;
+                    headers_[includedPath] = headers_.length;
+                }
+
+                auto pair = tuple(headers_[includedPath], headers_[path]);
+                directInclusions.add(pair);
+            }
+        }
+
+        foreach (header, closure; transitiveClosure())
+        {
+            foreach (closureHeader; closure)
+                indirectInclusions.add(tuple(header, closureHeader));
+        }
+    }
+
+    auto headers()
+    {
+        return headers_.byKey();
+    }
+
+    bool isIncludedBy(string header, string by)
+    {
+        return directInclusions.contains(
+            tuple(headers_[header.asAbsNormPath], headers_[by.asAbsNormPath]));
+    }
+
+    bool isReachableBy(string header, string by)
+    {
+        return indirectInclusions.contains(
+            tuple(headers_[header.asAbsNormPath], headers_[by.asAbsNormPath]));
+    }
+
+    bool isReachableThrough(string header, string via, string by)
+    {
+        return isReachableBy(header, via) && isReachableBy(via, by);
+    }
+
+    private size_t[] topologicalSort()
+    {
+        import std.range;
+
+        enum Status
+        {
+            pristine,
+            visiting,
+            visited
+        }
+
+        auto sorted = new size_t[0];
+        auto statuses = new Status[headers_.length];
+        auto includedBy = new size_t[][headers_.length];
+
+        foreach (inclusion; directInclusions.byKey)
+            includedBy[inclusion[0]] ~= inclusion[1];
+
+
+        void visit(size_t header)
+        {
+            if (statuses[header] == Status.visited)
+                return;
+
+            if (statuses[header] == Status.visiting)
+                throw new Exception("The include graph isn't a DAG.");
```

Comment:
Is this something that could happened or would this indicate a logical error in DStep?

Reply:
This would indicate logical error in dstep of libclang (it would happen if headers made cycle).